src/compiler/parser.ts

```diff
@@ -4150,8 +4150,15 @@ namespace ts {
                     return true;
                 }
 
+                // This handles cases like `await (async () => {})()`
+                // One drawback of this implementation is that we cannot use this form in top-level
+                // if there is an "await" identifier as we cannot differentiate this two cases
+                const isCallLike = () => {
+                    return (nextToken() === SyntaxKind.OpenParenToken && !identifiers.has("await")) && !scanner.hasPrecedingLineBreak();

```

Comment:
The problem with this approach is function hoisting. Your example has this:
```ts
function await(fn) {
  return () => {};
}
await (async () => {})(); // equivalent to: await((async () => {})())
```
However, it should also fail in this case:
```ts
await (async () => {})(); // equivalent to: await((async () => {})())
function await(fn) {
  return () => {};
}
```
I don't think this is the correct solution to the underlying problem.